### PR TITLE
[mempool] Allow one extra single-ancestor transaction per package

### PR DESCRIPTION
**Pull request by TheBlueMatt**
This implements the proposed policy change from [1], which allows
certain classes of contract protocols involving revocation
punishments to use CPFP. Note that some such use-cases may still
want some form of one-deep package relay, though even this alone
may greatly simplify some lightning fee negotiation.

[1] https://lists.linuxfoundation.org/pipermail/bitcoin-dev/2018-November/016518.html

### CODE REVIEW

File: src/validation.cpp [hunk 1]
```diff
@@ -761,7 +761,14 @@ static bool AcceptToMemoryPoolWorker(const CChainParams& chainparams, CTxMemPool
         size_t nLimitDescendantSize = gArgs.GetArg("-limitdescendantsize", DEFAULT_DESCENDANT_SIZE_LIMIT)*1000;
         std::string errString;
         if (!pool.CalculateMemPoolAncestors(entry, setAncestors, nLimitAncestors, nLimitAncestorSize, nLimitDescendants, nLimitDescendantSize, errString)) {
-            return state.DoS(0, false, REJECT_NONSTANDARD, "too-long-mempool-chain", false, errString);
+            setAncestors.clear();
+            // If the new transaction is relatively small (up to 40k weight)
+            // and has at most one ancestor (ie ancestor limit of 2, including
+            // the new transaction), allow it if its parent has exactly the
+            // descendant limit descendants.
+            if (nSize <= 10000 && !pool.CalculateMemPoolAncestors(entry, setAncestors, 2, 1000000, nLimitDescendants + 1, nLimitDescendantSize + 10000, errString)) {
```
Comment (instagibbs) [inline src/validation.cpp]: I'm still chewing on the comment above, but doesn't this line let transactions >40k weight through no matter what?

Comment (TheBlueMatt) [reply src/validation.cpp]: Lol, oops, one of these days I'll learn to code.

Comment (instagibbs) [reply src/validation.cpp]: did the tests work? ;D

File: test/functional/mempool_package_onemore.py [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+        assert_raises_rpc_error(-26, "too-long-mempool-chain", self.chain_transaction, self.nodes[0], [txid], [0], value, fee, 1)
+        # ...even if it chains on from some point in the middle of the chain.
+        assert_raises_rpc_error(-26, "too-long-mempool-chain", self.chain_transaction, self.nodes[0], [chain[2][0]], [1], chain[2][1], fee, 1)
+        assert_raises_rpc_error(-26, "too-long-mempool-chain", self.chain_transaction, self.nodes[0], [chain[1][0]], [1], chain[1][1], fee, 1)
+        # ...even if it chains on to two parent transactions with one in the chain.
+        assert_raises_rpc_error(-26, "too-long-mempool-chain", self.chain_transaction, self.nodes[0], [chain[0][0], second_chain], [1, 0], chain[0][1] + second_chain_value, fee, 1)
+        # But not if it chains directly off the first transaction
+        self.chain_transaction(self.nodes[0], [chain[0][0]], [1], chain[0][1], fee, 1)
+        # and the second chain should work just fine
+        self.chain_transaction(self.nodes[0], [second_chain], [0], second_chain_value, fee, 1)
```
Comment (instagibbs) [inline test/functional/mempool_package_onemore.py]: Add a test that blowing the 40k weight makes it fail?

Comment (TheBlueMatt) [reply test/functional/mempool_package_onemore.py]: Done!

**Comment by TheBlueMatt** [comment]
CC @rustyrussell.

File: test/functional/mempool_package_onemore.py [hunk 3]
```diff
@@ -0,0 +1,87 @@
+#!/usr/bin/env python3
+# Copyright (c) 2014-2019 The Bitcoin Core developers
+# Distributed under the MIT software license, see the accompanying
+# file COPYING or http://www.opensource.org/licenses/mit-license.php.
+"""Test descendant package tracking carge-out allowing one final transaction in
```
Comment (practicalswift) [inline test/functional/mempool_package_onemore.py]: Should be "carve" :-)

File: test/functional/mempool_package_onemore.py [hunk 4]
```diff
@@ -0,0 +1,87 @@
...
+    def set_test_params(self):
+        self.num_nodes = 2
+        self.extra_args = [["-maxorphantx=1000"], ["-maxorphantx=1000", "-limitancestorcount=5"]]
+
+    def skip_test_if_missing_module(self):
+        self.skip_if_no_wallet()
+
+    # Build a transaction that spends parent_txid:vout
+    # Return amount sent
+    def chain_transaction(self, node, parent_txids, vouts, value, fee, num_outputs):
```
Comment (practicalswift) [inline test/functional/mempool_package_onemore.py]: Nit: This method doesn't use `self`: could be made a function?

File: test/functional/mempool_package_onemore.py [hunk 5]
```diff
@@ -0,0 +1,87 @@
...
+        send_value = satoshi_round((value - fee)/num_outputs)
+        inputs = []
+        for (txid, vout) in zip(parent_txids, vouts):
+            inputs.append({'txid' : txid, 'vout' : vout})
+        outputs = {}
+        for i in range(num_outputs):
```
Comment (practicalswift) [inline test/functional/mempool_package_onemore.py]: Nit: `for _` is more idiomatic when `_` is unused :-)  Applies also for the other `for i`:s below :-)

File: src/validation.cpp [hunk 6]
```diff
@@ -761,7 +761,14 @@ static bool AcceptToMemoryPoolWorker(const CChainParams& chainparams, CTxMemPool
         size_t nLimitDescendantSize = gArgs.GetArg("-limitdescendantsize", DEFAULT_DESCENDANT_SIZE_LIMIT)*1000;
         std::string errString;
         if (!pool.CalculateMemPoolAncestors(entry, setAncestors, nLimitAncestors, nLimitAncestorSize, nLimitDescendants, nLimitDescendantSize, errString)) {
-            return state.DoS(0, false, REJECT_NONSTANDARD, "too-long-mempool-chain", false, errString);
+            setAncestors.clear();
+            // If the new transaction is relatively small (up to 40k weight)
+            // and has at most one ancestor (ie ancestor limit of 2, including
+            // the new transaction), allow it if its parent has exactly the
+            // descendant limit descendants.
```
Comment (morcos) [inline src/validation.cpp]: Please expand significantly on the rationale for this change because its a weird little bit of added complication, and basically the only downside is people not understanding why it's there.  You could link to the mailing list post maybe.

File: src/validation.cpp (hunk 1 above)
Comment (jachiang) [reply src/validation.cpp]: @TheBlueMatt I have a spill-over question from the PR review club which covered this PR last week. I am unsure of why the LimitAncestorSize for the carve-out tx has been increased to the block-size. 

If the parent of the carve-out tx were close to the LimitAncestorSize, this would create space for the carve-out tx. Preceding (low-fee-rate) descendent chains of the (carve-out) parent would have to adhere to LimitAncestorSize I believe. 

However, what if the first parent descendent chain was from an actor intending to raise package fee-rate? This (well-intentioned) descendent chain would be burdened by LimitAncestorSize, in contrast to the subsequently broadcasted carve-out transaction, right? In an extreme case, the first descendent chain may even have less than 40k weight to work with?

How should one best reason about this? Thanks.

Comment (TheBlueMatt) [reply src/validation.cpp]: Indeed, we don't really care about the size of the ancestor transaction here (its limited to the standard single-tx size limit, or 100k today), so I just set it to 1M, we already check that there is only, at max, one unconfirmed parent.

**Comment by TheBlueMatt** [comment]
Resolved comments.

**Review by rustyrussell — APPROVED** [review]
Concept Ack.

After much discussion, this does seem to be the minimal-intervention fix.  Thanks Matt.

**Comment by sdaftuar** [comment]
Concept ACK.

I did some review and light testing, and I noticed that there seems to be a general problem with our RBF logic: if we have some parent transaction whose descendant count is maxed out in the mempool, then we're unable to rbf any child transaction, because we evaluate the descendant limits of a new transactions ancestors prior to looking at what might be evicted by that transaction.  (However, we can generally RBF the parent itself!)

So in particular, that means that the new transaction that would be carved out by this policy would not be able to be RBF'ed in the situation that the parent transactions' descendant limit has been reached.

I'm guessing we may want to fix that to provide additional footgun protection to users that would take advantage of this new behavior, though I'll leave it to the PR author to decide if it is worth doing in this PR or separately.

**Comment by ajtowns** [comment]
Concept ACK. Might be worth adding an explicit test case for RBF case @sdaftuar points out:

             # But not if it chains directly off the first transaction
             self.chain_transaction(self.nodes[0], [chain[0][0]], [1], chain[0][1], fee, 1)
    +
    +        # Sadly, RBFing that should fail, however
    +        assert_raises_rpc_error(-26, "too-long-mempool-chain", self.chain_transaction, 0, [chain[0][0]], [1], chain[0][1], fee*2, 1)
    +
             # and the second chain should work just fine
             self.chain_transaction(self.nodes[0], [second_chain], [0], second_chain_value, fee, 1)

(means `createrawtransaction` needs to be called with `replaceable=True` as well)

Could change `chain_transaction(self, node, ...)` to always use `node=self.nodes[0]` which would simplify calls, and deal with @practicalswift's nit.

Could define a const for the 10k magic number. Is there any reason to have 1M instead of reusing `nLimitAncestorSize`? (Could not allow the carve out at all if `nLimitAncestors <= 1` though that's probably getting a bit ridiculous)

I think this should work for two-party lightning channels, because the outputs will be a bunch of CSV-limited ones that won't be in the mempool, and one CPFP output for each party. So at most one output can be spammed, and that only up to the descendent limit, so the plus-one here should work fine. If there were three or more immediately-spendable outputs (for channel factories and multiparty eltoo eg), I don't think this would be enough. But that's fine today.

I guess it's not very useful to log when this carve out triggers -- it's meant to prevent attacks from being effective, so while it's there there won't be any attacks so it'll hardly ever be used, but if it weren't there there would be attacks and it would be useful.

Should this carve out be limited to only apply when it's actually raising the parent's effective fee rate? I think you'd have to loop over all the descendents of your parent, work out the best ancestor fee rate of any of them, then see if the new tx and the parent have a better combined fee rate than that. That seems like it would add a chunk of complexity though, so probably isn't worth it.

This seems like it's a sufficient workaround for current problems, and is a pretty minimal change, while doing anything much better would be lots more complicated. So: Approach ACK.

File: test/functional/mempool_package_onemore.py [hunk 7]
```diff
@@ -0,0 +1,87 @@
...
+from test_framework.test_framework import BitcoinTestFramework
+from test_framework.util import assert_equal, assert_raises_rpc_error, satoshi_round
+
+MAX_ANCESTORS = 25
+MAX_DESCENDANTS = 25
+
+class MempoolPackagesTest(BitcoinTestFramework):
+    def set_test_params(self):
+        self.num_nodes = 2
+        self.extra_args = [["-maxorphantx=1000"], ["-maxorphantx=1000", "-limitancestorcount=5"]]
```
Comment (ajtowns) [inline test/functional/mempool_package_onemore.py]: You're only ever using node 0 in this test

**Review by ryanofsky — APPROVED** [review]
ACK f1facdd3e27d3b5c0536ab27d1f928984b8ccc25 with minimal testing. I verified new test fails without policy change and succeeds with it. I left only minor review comments which can be ignored.

I do like ajtowns's RBF test from https://github.com/bitcoin/bitcoin/pull/15681#issuecomment-509146925, and think it'd be nice if that were added.

File: test/functional/mempool_package_onemore.py [hunk 8]
```diff
@@ -0,0 +1,87 @@
...
+        (second_chain, second_chain_value) = self.chain_transaction(self.nodes[0], [utxo[1]['txid']], [utxo[1]['vout']], utxo[1]['amount'], fee, 1)
+
+        # Check mempool has MAX_ANCESTORS transactions in it, and descendant and ancestor
+        # count and fees should look correct
+        assert_equal(len(self.nodes[0].getrawmempool(True)), MAX_ANCESTORS + 1)
+
+        # Adding one more transaction on to the chain should fail.
+        assert_raises_rpc_error(-26, "too-long-mempool-chain", self.chain_transaction, self.nodes[0], [txid], [0], value, fee, 1)
+        # ...even if it chains on from some point in the middle of the chain.
+        assert_raises_rpc_error(-26, "too-long-mempool-chain", self.chain_transaction, self.nodes[0], [chain[2][0]], [1], chain[2][1], fee, 1)
```
[the remark this answers is not in the document]
Comment (ajtowns) [reply test/functional/mempool_package_onemore.py]: FWIW, I thought the way it is seemed more readable. YMMV obviously :)

File: test/functional/mempool_package_onemore.py [hunk 9]
```diff
@@ -0,0 +1,87 @@
...
+
+        fee = Decimal("0.0002")
+        # MAX_ANCESTORS transactions off a confirmed tx should be fine
+        chain = []
+        for _ in range(4):
+            (txid, sent_value) = self.chain_transaction(self.nodes[0], [txid], [vout], value, fee, 2)
+            vout = 0
+            value = sent_value
+            chain.append([txid, value])
+        for _ in range(MAX_ANCESTORS - 4):
```
[the remark this answers is not in the document]
Comment (TheBlueMatt) [reply test/functional/mempool_package_onemore.py]: I find this more readable.

File: test/functional/mempool_package_onemore.py (hunk 7 above)
Comment (TheBlueMatt) [reply test/functional/mempool_package_onemore.py]: Fixed.

**Comment by TheBlueMatt** [comment]
@ajtowns I'd prefer to not add the proposed test change, as @sdaftuar's comment points out a critical limitation, and one I'd like to get resolved before this gets released, though not necessarily in this PR. I'd rather just fix the bug than add a test to ensure that the bug exists.

As for general policy questions: this doesn't result in any new transactions being accepted which violate min fee, nor does it change eviction criteria when mempool fills. What this *does* do, is tweak our already-somewhat-arbitrary-but-sadly-neccessary anti-DoS policies to allow one more (potentially-in-the-future) common case so that we can maximize fee revenue in the case this does get used. Thus, I don't see any reason to need to bend over backwards to make it more restrictive, if anything, less restrictive is better.

**Comment by sdaftuar** [comment]
ACK f1facdd3e27d3b5c0536ab27d1f928984b8ccc25

**Review by ryanofsky — APPROVED** [review]
utACK 50cede3f5a4d4fbfbb7c420b94e661a6a159bced. Changes since last review: adding EXTRA_DESCENDANT_TX_SIZE_LIMIT constant, changing max ancestor size from 1,000,000 to nLimitAncestorSize constant (101,000), fixing test comment and getting rid of unused test node.

**Comment by ajtowns** [comment]
ACK 50cede3f5a4d4fbfbb7c420b94e661a6a159bced -- looked over code again, compared with previous commit, compiles, etc.

I think this is a sufficient improvement even with the problem @sdaftuar points out:

 * if there's no adversarial situation and you hit the descendent limit with low fee children just due to carelessness, this carveout can be used to RBF the first child (paying 25 times the feerate to account for the descendants' fees) and make CPFP actually work in ways where you can't do anything now
 * in the event of an adversarial situation with only two immediately spendable outputs, if you do the CPFP first, they can't block you from RBF'ing when you need to (they can't make use of the onemore carve out if they only have one output available -- they need to use it to hit the descant limit, at which point it's not available for onemore).
 * to DoS their channel partner, an adversary would need to be confident their partner will CPFP by too small an amount, and do their spam low-fee tx spamming before they do that. If the CPFP is a high enough amount, they'll likely lose the fees for their spam once the mempool clears out, making that not a free attack, as well.

It would obviously still be better to find a more general fix, but this seems to me a worthwhile improvement without introducing much of a maintenance cost.

**Comment by ajtowns** [comment]
Patch to `mempool_package_onemore.py` for checking the "can RBF if I've added too many descendants myself" case. Maybe useful for reviewers, not suitable for inclusion since it kills the checks for the intended usecase:

```diff
+++ b/test/functional/mempool_package_onemore.py
@@ -33,7 +33,7 @@ class MempoolPackagesTest(BitcoinTestFramework):
         outputs = {}
         for i in range(num_outputs):
             outputs[node.getnewaddress()] = send_value
-        rawtx = node.createrawtransaction(inputs, outputs)
+        rawtx = node.createrawtransaction(inputs, outputs, 0, True)
         signedtx = node.signrawtransactionwithwallet(rawtx)
         txid = node.sendrawtransaction(signedtx['hex'])
         fulltx = node.getrawtransaction(txid, 1)
@@ -74,6 +74,12 @@ class MempoolPackagesTest(BitcoinTestFramework):
         assert_raises_rpc_error(-26, "too-long-mempool-chain", self.chain_transaction, self.nodes[0], [chain[0][0], second_chain], [1, 0], chain[0][1] + second_chain_value, fee, 1)
         # ...especially if its > 40k weight
         assert_raises_rpc_error(-26, "too-long-mempool-chain", self.chain_transaction, self.nodes[0], [chain[0][0]], [1], chain[0][1], fee, 350)
+
+        # can even RBF the original tx
+        self.chain_transaction(self.nodes[0], [chain[0][0]], [0], chain[0][1], fee*25, 10)
+        assert_equal(len(self.nodes[0].getrawmempool(True)), 3) # parent, second_chain, RBF'd original
+        return
+
         # But not if it chains directly off the first transaction
         self.chain_transaction(self.nodes[0], [chain[0][0]], [1], chain[0][1], fee, 1)
         # and the second chain should work just fine
```

**Comment by sdaftuar** [comment]
ACK 50cede3f5a4d4fbfbb7c420b94e661a6a159bced

File: src/validation.cpp [hunk 10]
```diff
@@ -610,7 +610,21 @@ static bool AcceptToMemoryPoolWorker(const CChainParams& chainparams, CTxMemPool
...
+            //
+            // This allows protocols which rely on distrusting counterparties
+            // being able to broadcast descendants of an unconfirmed transaction
+            // to be secure by simply only having two immediately-spendable
+            // outputs - one for each counterparty. For more info on the uses for
+            // this, see https://lists.linuxfoundation.org/pipermail/bitcoin-dev/2018-November/016518.html
+            if (nSize >  EXTRA_DESCENDANT_TX_SIZE_LIMIT ||
+                    !pool.CalculateMemPoolAncestors(entry, setAncestors, 2, nLimitAncestorSize, nLimitDescendants + 1, nLimitDescendantSize + EXTRA_DESCENDANT_TX_SIZE_LIMIT, errString)) {
```
Comment (jnewbery) [inline src/validation.cpp]: Passing `errString` through to the second call of `CalculateMemPoolAncestors()` means that the error string in the `CValidationState` object will be incorrect because of the changed ancestor/descendant size/count parameters. Really we should declare a `dummy_error_string` to pass into the second call, and use the original `errString` in the call to `state.Invalid()` below.

File: test/functional/mempool_package_onemore.py [hunk 11]
```diff
@@ -0,0 +1,86 @@
+#!/usr/bin/env python3
+# Copyright (c) 2014-2019 The Bitcoin Core developers
+# Distributed under the MIT software license, see the accompanying
+# file COPYING or http://www.opensource.org/licenses/mit-license.php.
+"""Test descendant package tracking carve-out allowing one final transaction in
+   an otherwise-full package as long as it has only one parent and is <= 10k in
+   size.
```
Comment (jnewbery) [inline test/functional/mempool_package_onemore.py]: nit: virtual size (or define this in terms or transaction weight)

File: test/functional/mempool_package_onemore.py [hunk 12]
```diff
@@ -0,0 +1,86 @@
...
+        signedtx = node.signrawtransactionwithwallet(rawtx)
+        txid = node.sendrawtransaction(signedtx['hex'])
+        fulltx = node.getrawtransaction(txid, 1)
+        assert len(fulltx['vout']) == num_outputs  # make sure we didn't generate a change output
+        return (txid, send_value)
+
+    def run_test(self):
+        # Mine some blocks and have them mature.
+        self.nodes[0].generate(101)
+        utxo = self.nodes[0].listunspent(10)
```
Comment (jnewbery) [inline test/functional/mempool_package_onemore.py]: nit: it's preferable to use named transactions to make the test more readable.

File: test/functional/mempool_package_onemore.py [hunk 13]
```diff
@@ -0,0 +1,86 @@
...
+
+from decimal import Decimal
+
+from test_framework.test_framework import BitcoinTestFramework
+from test_framework.util import assert_equal, assert_raises_rpc_error, satoshi_round
+
+MAX_ANCESTORS = 25
+MAX_DESCENDANTS = 25
+
+class MempoolPackagesTest(BitcoinTestFramework):
```
Comment (jnewbery) [inline test/functional/mempool_package_onemore.py]: nit: name this class `MempoolPackageOnemoreTest`

File: test/functional/mempool_package_onemore.py (hunk 4 above)
Comment (jnewbery) [reply test/functional/mempool_package_onemore.py]: agree that this would be better as a function than a method

File: test/functional/mempool_package_onemore.py (hunk 12 above)
Comment (jnewbery) [inline test/functional/mempool_package_onemore.py]: nit: name this variable `utxos` or define it as `self.nodes[0].listunspent(10)[0]`

File: test/functional/mempool_package_onemore.py [hunk 14]
```diff
@@ -0,0 +1,87 @@
...
+        chain = []
+        for _ in range(4):
+            (txid, sent_value) = self.chain_transaction(self.nodes[0], [txid], [vout], value, fee, 2)
+            vout = 0
+            value = sent_value
+            chain.append([txid, value])
+        for _ in range(MAX_ANCESTORS - 4):
+            (txid, sent_value) = self.chain_transaction(self.nodes[0], [txid], [0], value, fee, 1)
+            value = sent_value
+            chain.append([txid, value])
```
[the remark this answers is not in the document]
Comment (jnewbery) [reply test/functional/mempool_package_onemore.py]: +1, or used a namedtuple

File: test/functional/mempool_package_onemore.py (hunk 2 above)
Comment (jnewbery) [reply test/functional/mempool_package_onemore.py]: I don't think this addresses the edge case that we want. The line you added above:

`assert_raises_rpc_error(-26, "too-long-mempool-chain", self.chain_transaction, self.nodes[0], [chain[0][0]], [1], chain[0][1], fee, 350)`

Fails both the `nSize >  EXTRA_DESCENDANT_TX_SIZE_LIMIT` _and_ `!pool.CalculateMemPoolAncestors(entry, setAncestors, 2, nLimitAncestorSize, nLimitDescendants + 1, nLimitDescendantSize + EXTRA_DESCENDANT_TX_SIZE_LIMIT, errString)` checks. We want a test that differentitates just the weight > 40k case.

**Review by jnewbery — COMMENTED** [review]
Quick code-review ACK 50cede3f5a4d4fbfbb7c420b94e661a6a159bced.

There's one logging bug that I think we should fix. I've added a bunch of nits to the test, but I agree with other reviewers that we could probably avoid a bunch of test code duplication by merging `mempool_package_onemore.py` with `mempool_packages.py`.

**Comment by ysangkok** [comment]
just FYI: a lot of interesting comments on this PR available at: https://bitcoincore.reviews/15681.html
